File: 22_kth_node_to_tail/go/main.go
package main

import (
	"fmt"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func KthToTail(head *ListNode, k int) *ListNode {
	if head == nil || k <= 0 {
		return nil
	}
	front := head
	var back *ListNode
	for i := 0; i < k - 1; i++ {
		if front.Next == nil {
			return nil
		}
		front = front.Next
	}

	back = head
	for front.Next != nil {
		front, back = front.Next, back.Next
	}
	return back
}

func main() {
	node1 := &ListNode{1, nil}
	node2 := &ListNode{2, nil}
	node1.Next = node2
	node3 := &ListNode{3, nil}
	node2.Next = node3
	node4 := &ListNode{4, nil}
	node3.Next = node4
	node5 := &ListNode{5, nil}
	node4.Next = node5
	node6 := &ListNode{6, nil}
	node5.Next = node6
	node7 := &ListNode{7, nil}
	node6.Next = node7

	fmt.Println(KthToTail(node1, 7).Val)
}

File: 22_kth_node_to_tail/main.py
# -*- coding: utf-8 -*-


class ListNode(object):

    def __init__(self, val):
        self.val = val
        self.next = None


def kth_to_tail(head, k):
    if head is None or k <= 0:
        return None
    
    front, back = head, None
    for i in range(k - 1):
        if front.next is None:
            return None
        front = front.next

    back = head
    while front.next is not None:
        front = front.next
        back = back.next
    return back


if __name__  == '__main__':
    node1 = ListNode(1)
    node2 = ListNode(2)
    node1.next = node2
    node3 = ListNode(3)
    node2.next = node3
    node4 = ListNode(4)
    node3.next = node4
    node5 = ListNode(5)
    node4.next = node5
    node6 = ListNode(6)
    node5.next = node6
    node7 = ListNode(7)
    node6.next = node7

    print(kth_to_tail(node1, 3).val)
